advancedstats.py

```diff
@@ -0,0 +1,187 @@
+import pandas as pd
+
+pd.set_option('display.max_rows', 500)
+pd.set_option('display.max_columns', 500)
+pd.set_option('display.width', 1000)
+
+def prepare_data(detailed_season_results_path):
+	'''
+	Returns dataframe that is suitable for performing group by.
+	It cleans any data to make it suitable for further processing
+	:param detailed_season_results_path:
+	:return:
+	'''
+	# Load the season results
+	seasons_results = pd.read_csv(detailed_season_results_path)
+
+	# Create a copy and rearrange column data
+	seasons_results_copy = seasons_results.copy()
+	seasons_results_copy[['WTeamID', 'WScore', 'WFGM', 'WFGA','WFGM3', 'WFGA3', 'WFTM', 'WFTA',	'WOR', 'WDR', 'WAst', 'WTO', 'WStl', 'WBlk', 'WPF']] = seasons_results[['LTeamID', 'LScore', 'LFGM', 'LFGA','LFGM3', 'LFGA3', 'LFTM', 'LFTA', 'LOR', 'LDR', 'LAst', 'LTO', 'LStl', 'LBlk', 'LPF']]
+	seasons_results_copy[
+		['LTeamID', 'LScore', 'LFGM', 'LFGA', 'LFGM3', 'LFGA3', 'LFTM', 'LFTA', 'LOR', 'LDR', 'LAst', 'LTO', 'LStl',
+		 'LBlk', 'LPF']] = seasons_results[
+		['WTeamID', 'WScore', 'WFGM', 'WFGA', 'WFGM3', 'WFGA3', 'WFTM', 'WFTA', 'WOR', 'WDR', 'WAst', 'WTO', 'WStl',
+		 'WBlk', 'WPF']]
+
+	seasons_results_concat = pd.concat([seasons_results, seasons_results_copy], ignore_index=True)
+
+	# Rename the columns
+	seasons_results_concat.rename(columns={
+		'WTeamID': 'TeamID_1',
+		'WScore': 'Score_1',
+		'LTeamID': 'TeamID_2',
+		'LScore': 'Score_2',
+		# Winning team
+		'WFGM': 'FGM_1',
+		'WFGA': 'FGA_1',
+		'WFGM3': 'FGM3_1',
+		'WFGA3': 'FGA3_1',
+		'WFTM': 'FTM_1',
+		'WFTA': 'FTA_1',
+		'WOR': 'OR_1',
+		'WDR': 'DR_1',
+		'WAst': 'Ast_1',
+		'WTO': 'TO_1',
+		'WStl': 'Stl_1',
+		'WBlk': 'Blk_1',
+		'WPF': 'PF_1',
+		# Losing team
+		'LFGM': 'FGM_2',
+		'LFGA': 'FGA_2',
+		'LFGM3': 'FGM3_2',
+		'LFGA3': 'FGA3_2',
+		'LFTM': 'FTM_2',
+		'LFTA': 'FTA_2',
+		'LOR': 'OR_2',
+		'LDR': 'DR_2',
+		'LAst': 'Ast_2',
+		'LTO': 'TO_2',
+		'LStl': 'Stl_2',
+		'LBlk': 'Blk_2',
+		'LPF': 'PF_2'
+	}, inplace=True)
+
+	return seasons_results_concat
+
+def get_poss(tm_FGA, tm_FGM, tm_FTA, tm_OR, tm_DR, tm_TO, opp_FGA, opp_FGM, opp_FTA, opp_OR, opp_DR, opp_TO):
+	tm_half = tm_FGA + 0.4*tm_FTA - 1.07*(tm_OR / (tm_OR + opp_DR))*(tm_FGA - tm_FGM) + tm_TO
+	opp_half = opp_FGA + 0.4*opp_FTA - 1.07*(opp_OR / (opp_OR + tm_DR))*(opp_FGA - opp_FGM) + opp_TO
+	return 0.5*(tm_half+opp_half)
+
+def get_ORtg(points, poss):
+	return 100 * points / poss
+
+def get_DRtg(opp_points_allowed, poss):
+	return 100 * opp_points_allowed / poss
+
+def get_season_totals(data):
+	season_team_totals = data.groupby(['Season', 'TeamID_1']).agg({
```

Comment:
think this can be simplify to doing one groupby then sum first
```
data.groupby(['Season', 'TeamID_1').sum().reset_index()
```

then if we want game count for team + season, we can do

```
game_counts = data.groupby(['Season', 'TeamID_1').size().reset_index(name='counts')
```

and then merge the two dataframe into one

Reply:
What does the `reset_index(name='counts')` do?

Reply:
it set the column with the game count with the column name of `count`